src/mscorlib/shared/System/ReadOnlyMemory.cs

```diff
@@ -0,0 +1,295 @@
+        /// <param name="length">The desired length for the slice (exclusive).</param>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the specified <paramref name="start"/> or end index is not in range (&lt;0 or &gt;=Length).
+        /// </exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public ReadOnlyMemory<T> Slice(int start, int length)
+        {
+            if ((uint)start > (uint)_length || (uint)length > (uint)(_length - start))
+                ThrowHelper.ThrowArgumentOutOfRangeException();
+
+            // There is no need to 'and' _index by the bit mask here 
+            // since the constructor will set the highest order bit again anyway
+            if (_index < 0)
+                return new ReadOnlyMemory<T>(Unsafe.As<OwnedMemory<T>>(_arrayOrOwnedMemory), _index + start, length);
+            return new ReadOnlyMemory<T>(Unsafe.As<T[]>(_arrayOrOwnedMemory), _index + start, length);
+        }
+
+        /// <summary>
+        /// Returns a span from the memory.
+        /// </summary>
+        public ReadOnlySpan<T> Span
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get
+            {
+                if (_index < 0)
+                    return Unsafe.As<OwnedMemory<T>>(_arrayOrOwnedMemory).AsSpan(_index & bitMask, _length);
+                return new ReadOnlySpan<T>(Unsafe.As<T[]>(_arrayOrOwnedMemory), _index, _length);
+            }
+        }
+
+        public unsafe MemoryHandle Retain(bool pin = false)
+        {
+            MemoryHandle memoryHandle;
+            if (pin)
+            {
+                if (_index < 0)
+                {
+                    memoryHandle = Unsafe.As<OwnedMemory<T>>(_arrayOrOwnedMemory).Pin();
+                }
+                else
+                {
+                    var handle = GCHandle.Alloc(Unsafe.As<T[]>(_arrayOrOwnedMemory), GCHandleType.Pinned);
+                    void* pointer = Unsafe.Add<T>((void*)handle.AddrOfPinnedObject(), _index);
+                    memoryHandle = new MemoryHandle(null, pointer, handle);
+                }
+            }
+            else
+            {
+                if (_index < 0)
+                {
+                    Unsafe.As<OwnedMemory<T>>(_arrayOrOwnedMemory).Retain();
+                    memoryHandle = new MemoryHandle(Unsafe.As<OwnedMemory<T>>(_arrayOrOwnedMemory));
+                }
+                else
+                {
+                    memoryHandle = new MemoryHandle(null);
+                }
+            }
+            return memoryHandle;
+        }
+
+        /// <summary>
+        /// Get an array segment from the underlying memory. 
+        /// If unable to get the array segment, return false with a default array segment.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public bool DangerousTryGetArray(out ArraySegment<T> arraySegment)
```

Comment:
We call it Dangerous here because it gives access that's not denoted by the type name (you can get the array to write to even though the type is called "read only").  But a similar issue exists on `Memory<T>`: the `Memory<T>` could have been constructed for a specific range of an array, but calling `TryGetArray` gives you back an `ArraySegment<T>`, which exposes the `Array`, and thus potentially gives you access to portions of data that you weren't supposed to.  Do we care?  I'm wondering if that one should be called `Dangerous`, too, or maybe we don't care about that distinction?

Reply:
Do you suggest we mark `TryGetArray` with the [EditorBrowsable(EditorBrowsableState.Never)] attribute as well (along with calling it Dangerous)?

Reply:
No, I'd leave it visible.  And to be clear, I also wasn't pushing for Memory.TryGetArray to be renamed, rather making sure we though through whether it should be or not.

Reply:
I think ArraySegment already communicates that array items outside of the segment's range should not be accessed.